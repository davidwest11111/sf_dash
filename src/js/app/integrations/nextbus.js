import axios from 'axios';
import X2JS from 'x2js';
import qs from 'qs';


export const XML2JS_CONFIG = {
  arrayAccessFormPaths: [
    'body.predictions',
    'body.predictions.direction.prediction',
  ],
};
export const URL = 'http://webservices.nextbus.com/service/publicXMLFeed';

export default class NextBus {
  static getPredictions = async stopRouteTags => {
    const response = await axios({
      method: 'get',
      url: URL,
      params: {
        command: 'predictionsForMultiStops',
        a: 'sf-muni',
        stops: stopRouteTags,
        useShortTitles: true,
      },
      paramsSerializer: params => qs.stringify(params, { arrayFormat: 'repeat' }),
    });
    return new X2JS(XML2JS_CONFIG).xml2js(response.data);
  }

  static getRouteConfig = async locations => {
    // return $http.get(URL, {
    //   cache: true,
    //   params: {
    //     command: 'routeConfig',
    //     a: 'sf-muni',
    //     useShortTitles: true,
    //   },
    // })
    //   .then(data => this.mergeRouteData(data.data.body.route));
  }

  static getStopsWithin = async meters => {
    // const key = getLocalStorage(WEATHER_KEY_KEY);
    // const units = this.convertUnitsForDarkSky(getLocalStorage(WEATHER_UNITS_KEY));
    // const responses = await axios.all(
    //   locations.map(({ lat, long }) => (
    //     this.fetchWeatherData({ lat, long, key, units })
    //   )),
    // );
    // return responses;


    // return $q.all([
    //   geolocation.getLocation(),
    //   this.getRouteConfig(),
    // ]).then(promises => {
    //   /** Once we've got both the user's location and the route Data search */
    //   const stopResults = [];
    //   const currPosition = new google.maps.LatLng(
    //     promises[0].coords.latitude,
    //     promises[0].coords.longitude,
    //   );
    //   const routeConfig = promises[1];
    //
    //   routeConfig.forEach(route => {
    //     route.direction.forEach(direction => {
    //       direction.stop.forEach(stop => {
    //         /** Find stops that are within [meters] of the users position */
    //         const stopLatLng = new google.maps.LatLng(stop._lat, stop._lon);
    //         const distance = google.maps.geometry.spherical
    //           .computeDistanceBetween(currPosition, stopLatLng);
    //         if (distance <= meters) {
    //           stopResults.push({
    //             tag: route._tag,
    //             title: route._title,
    //             directionName: direction._name,
    //             directionTitle: direction._title,
    //             latLng: stopLatLng,
    //             stopId: stop._stopId,
    //             stopTag: stop._tag,
    //             stopTitle: stop._title,
    //           });
    //         }
    //       });
    //     });
    //   });
    //
    //   return stopResults;
    // });
  }

  static mergeRouteData = routeData => {
    routeData.forEach(route => {
      // map the stop tags
      const stopTags = {};
      route.stop.forEach(stop => {
        stopTags[stop._tag] = stop;
      });

      route.direction.forEach(direction => {
        for (let i = 0; i < direction.stop.length; i++) { // eslint-disable-line no-plusplus
          const tag = direction.stop[i]._tag;
          direction.stop[i] = stopTags[tag];
        }
      });
    });
    return routeData;
  }
}
