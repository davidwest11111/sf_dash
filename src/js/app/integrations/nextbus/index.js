import axios from 'axios';
import X2JS from 'x2js';
import qs from 'qs';


export const URL = 'http://webservices.nextbus.com/service/publicXMLFeed';

export const getUserPosition = options => (
  new Promise((resolve, reject) => {
    navigator.geolocation.getCurrentPosition(resolve, reject, options);
  })
);

export default class NextBus {
  static async getPredictions(routeStopTags) {
    const response = await axios({
      method: 'get',
      url: URL,
      params: {
        command: 'predictionsForMultiStops',
        a: 'sf-muni',
        stops: routeStopTags,
        useShortTitles: true,
      },
      paramsSerializer: params => qs.stringify(params, { arrayFormat: 'repeat' }),
    });

    const data = this.parsePredictionsXMLResponse(response.data);

    return new Promise((resolve, reject) => {
      const errorText = data?.body?.Error?.__text;
      if (errorText) { reject(errorText); }

      resolve(data);
    });
  }

  static getRouteConfig = async () => {
    const response = await axios({
      method: 'get',
      url: URL,
      params: {
        command: 'routeConfig',
        a: 'sf-muni',
        useShortTitles: true,
      },
    });

    const data = this.parseRouteConfigXMLResponse(response.data);

    return new Promise((resolve, reject) => {
      const errorText = data?.body?.Error?.__text;
      if (errorText) { reject(errorText); }

      resolve(this.mergeRouteData(data.body.route));
    });
  }

  static async getStopsNearMe(meters) {
    const responses = await axios.all([
      getUserPosition(),
      this.getRouteConfig(),
    ]);

    const [geoLocation, routeConfig] = responses;

    const currPosition = new google.maps.LatLng(
      geoLocation.coords.latitude,
      geoLocation.coords.longitude,
    );

    // FIXME: this is old. see if i can't clean it up
    const stopResults = [];
    routeConfig.forEach(route => {
      route.direction.forEach(direction => {
        direction.stop.forEach(stop => {
          /** Find stops that are within [meters] of the users position */
          const stopLatLng = new google.maps.LatLng(stop._lat, stop._lon);
          const distance = google.maps.geometry.spherical
            .computeDistanceBetween(currPosition, stopLatLng);

          if (distance <= meters) {
            stopResults.push({
              tag: route._tag,
              title: route._title,
              directionName: direction._name,
              directionTitle: direction._title,
              latLng: stopLatLng,
              stopId: stop._stopId,
              stopTag: stop._tag,
              stopTitle: stop._title,
            });
          }
        });
      });
    });

    return stopResults;
  }

  static mergeRouteData = routeData => {
    // FIXME
    routeData.forEach(route => {
      // map the stop tags
      const stopTags = {};
      route.stop.forEach(stop => {
        stopTags[stop._tag] = stop;
      });

      route.direction.forEach(direction => {
        // FIXME
        for (let i = 0; i < direction.stop.length; i++) { // eslint-disable-line no-plusplus
          const tag = direction.stop[i]._tag;
          direction.stop[i] = stopTags[tag];
        }
      });
    });
    return routeData;
  }

  static async isValidStop(routeStopTag) {
    try {
      // FIXME: this.getPredictions doesn't actually return a Promise
      const response = await this.getPredictions([routeStopTag]);
      return !!response?.body?.predictions;
    } catch (e) {
      return false;
    }
  }

  static parsePredictionsXMLResponse(xmlString) {
    return new X2JS({
      arrayAccessFormPaths: [
        'body.predictions',
        'body.predictions.direction.prediction',
      ],
    }).xml2js(xmlString);
  }

  static parseRouteConfigXMLResponse(xmlString) {
    return new X2JS({
      arrayAccessFormPaths: [
        'body.route',
      ],
    }).xml2js(xmlString);
  }

  static getRouteStopTag(routeTag, stopTag) {
    return `${routeTag}|${stopTag}`;
  }
}
