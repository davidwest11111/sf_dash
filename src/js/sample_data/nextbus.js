// The data returned is actually in XML, this is just what the X2JS library outputs

export const SAMPLE_PREDICTION = {
  direction: {
    prediction: [
      {
        _epochTime: '1588198229803',
        _seconds: '296',
        _minutes: '4',
        _isDeparture: 'false',
        _dirTag: '14___I_F00',
        _vehicle: '6609',
        _block: '1620',
        _tripTag: '9359239',
        $$hashKey: 'object:19',
      },
      {
        _epochTime: '1588198681179',
        _seconds: '747',
        _minutes: '12',
        _isDeparture: 'false',
        _dirTag: '14___I_F00',
        _vehicle: '6599',
        _block: '1622',
        _tripTag: '9359240',
        $$hashKey: 'object:20',
      }, { _epochTime: '1588199412858', _seconds: '1479', _minutes: '24', _isDeparture: 'false', _dirTag: '14___I_F00', _vehicle: '6547', _block: '1633', _tripTag: '9359237', $$hashKey: 'object:21' }, { _epochTime: '1588199419316', _seconds: '1486', _minutes: '24', _isDeparture: 'false', _dirTag: '14___I_F00', _vehicle: '6533', _block: '1628', _tripTag: '9359243', $$hashKey: 'object:22' }],
    _title: 'Inbound to Ferry Plaza',
  },
  message: [
    {
      _text: 'Service frequency for this stop is 10 minutes or less.',
      _priority: 'Normal',
    },
    {
      _text: 'Muni for essential trips only. Masks or face coverings are required. Practice physical distancing. Visit sfmta.com/COVID19 for more info.',
      _priority: 'Low',
    }, { _text: 'Text COVID19SF to 888-777 for official updates. Visit sfdph.org', _priority: 'Low' }, { _text: 'Board thru rear doors. Front door for accessibility needs only.', _priority: 'Normal' }],
  _agencyTitle: 'SF Muni',
  _routeTitle: '14-Mission',
  _routeTag: '14',
  _stopTitle: 'Mission St & 24th St',
  _stopTag: '5565',
  $$hashKey: 'object:17',
  className: '',
};

// FIXME
export const SAMPLE_FAILED_PREDICTIONS = {

};

export const SAMPLE_PREDICTIONS_XML = `
<?xml version="1.0" encoding="utf-8" ?> 
<body copyright="All data copyright San Francisco Muni 2020.">
  <predictions agencyTitle="SF Muni" routeTitle="12-Folsom-Pacific" routeTag="12" stopTitle="Pacific Ave &amp; Grant Ave" stopTag="5837" dirTitleBecauseNoPredictions="Inbound to Van Ness">
    <message text="Muni for essential trips only. Masks or face coverings are required. Practice physical distancing. Visit sfmta.com/COVID19 for more info." priority="Low"/>
    <message text="Text COVID19SF to 888-777 for official updates. Visit sfdph.org" priority="Low"/>
  </predictions>
  <predictions agencyTitle="SF Muni" routeTitle="14-Mission" routeTag="14" stopTitle="Mission St &amp; 24th St" stopTag="5565">
    <direction title="Inbound to Ferry Plaza">
      <prediction epochTime="1588394902295" seconds="11" minutes="0" isDeparture="false" dirTag="14___I_F00" vehicle="6591" block="1631" tripTag="9359290" />
      <prediction epochTime="1588395526282" seconds="635" minutes="10" isDeparture="false" dirTag="14___I_F00" vehicle="6539" block="1633" tripTag="9359291" />
    </direction>
    <message text="Service frequency for this stop is 10 minutes or less." priority="Normal"/>
    <message text="Muni for essential trips only. Masks or face coverings are required. Practice physical distancing. Visit sfmta.com/COVID19 for more info." priority="Low"/>
    <message text="Text COVID19SF to 888-777 for official updates. Visit sfdph.org" priority="Low"/>
    <message text="Board thru rear doors. Front door for accessibility needs only." priority="Normal"/>
  </predictions>
</body>
`;
