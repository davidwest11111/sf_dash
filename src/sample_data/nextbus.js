// The data returned is actually in XML, this is just what the X2JS library outputs

export const SAMPLE_PREDICTION = {
  direction: {
    prediction: [
      {
        _epochTime: '1588198229803',
        _seconds: '296',
        _minutes: '4',
        _isDeparture: 'false',
        _dirTag: '14___I_F00',
        _vehicle: '6609',
        _block: '1620',
        _tripTag: '9359239',
        $$hashKey: 'object:19',
      },
      {
        _epochTime: '1588198681179',
        _seconds: '747',
        _minutes: '12',
        _isDeparture: 'false',
        _dirTag: '14___I_F00',
        _vehicle: '6599',
        _block: '1622',
        _tripTag: '9359240',
        $$hashKey: 'object:20',
      }, { _epochTime: '1588199412858', _seconds: '1479', _minutes: '24', _isDeparture: 'false', _dirTag: '14___I_F00', _vehicle: '6547', _block: '1633', _tripTag: '9359237', $$hashKey: 'object:21' }, { _epochTime: '1588199419316', _seconds: '1486', _minutes: '24', _isDeparture: 'false', _dirTag: '14___I_F00', _vehicle: '6533', _block: '1628', _tripTag: '9359243', $$hashKey: 'object:22' }],
    _title: 'Inbound to Ferry Plaza',
  },
  message: [
    {
      _text: 'Service frequency for this stop is 10 minutes or less.',
      _priority: 'Normal',
    },
    {
      _text: 'Muni for essential trips only. Masks or face coverings are required. Practice physical distancing. Visit sfmta.com/COVID19 for more info.',
      _priority: 'Low',
    }, { _text: 'Text COVID19SF to 888-777 for official updates. Visit sfdph.org', _priority: 'Low' }, { _text: 'Board thru rear doors. Front door for accessibility needs only.', _priority: 'Normal' }],
  _agencyTitle: 'SF Muni',
  _routeTitle: '14-Mission',
  _routeTag: '14',
  _stopTitle: 'Mission St & 24th St',
  _stopTag: '5565',
  $$hashKey: 'object:17',
  className: '',
};

// FIXME
export const SAMPLE_FAILED_PREDICTIONS = {

};
