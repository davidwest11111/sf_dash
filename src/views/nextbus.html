<div class="nextbus">
  <div ng-show="!nextBus.predictions.length" class="card mb-2">
    <div class="card-body">
      No bus stops added.
    </div>
  </div>

  <div class="predictions" ng-include="'views/nextbus/predictions.html'"></div>

  <div bs-tabs class="add-form" ng-show="nextBus.addForm.showBusAddForm">
    <div class="btn close-form" ng-click="nextBus.addForm.toggleBusAddForm()"><span class="fas fa-times"></span></div>
    <div data-title="Geolocation" ng-include="'views/nextbus/add-stop-geolocate.html'" bs-pane></div>
    <div data-title="Manual Input" ng-include="'views/nextbus/add-stop-manual.html'" bs-pane></div>
  </div>

  <div ng-hide="nextBus.addForm.showBusAddForm" class="d-flex align-items-start justify-content-between">
    <div>
      <span ng-hide="nextBus.showBusRemoval" class="btn btn-xs btn-outline-secondary fas fa-plus" ng-click="nextBus.addForm.toggleBusAddForm()"></span>
      <span class="btn btn-xs btn-outline-secondary fas fa-pen" ng-hide="!nextBus.predictions.length" ng-click="nextBus.toggleBusRemove()"></span>
    </div>

    <react-component name="TimeSinceLastUpdated" props="nextBus" />
  </div>
</div>
