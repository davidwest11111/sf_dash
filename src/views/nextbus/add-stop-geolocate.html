<div class="card">
  <div class="card-body">
    <div class="alert alert-danger">
      Warning: This probably won't work. See
      <a
        href="https://developers.google.com/web/updates/2016/04/geolocation-on-secure-contexts-only"
        target="_blank"
        rel="noopener noreferrer"
      >
        Geolocation API Removed from Unsecured Origins in Chrome 50
      </a>.
    </div>

    <form class="form-inline">
      <span class="mr-2">Find stops within</span>
      <select class="form-control" ng-change="nextBus.addForm.getNearbyStops()" ng-model="nextBus.addForm.distance">
        <option value="">--Choose One--</option>
        <option value="322">1/5 mile</option>
        <option value="402">1/4 mile</option>
        <option value="804">1/2 mile</option>
        <!-- values are in meters -->
      </select>
    </form>
    <div class="spinner-loader" ng-show="nextBus.addForm.loading"></div>
    <div class="text-danger" ng-show="nextBus.addForm.errMsg">{{nextBus.addForm.errMsg}}</div>

    <table class="table table-striped table-hover" ng-show="nextBus.addForm.nearbyStops">
      <thead>
        <tr>
          <th>Route</th>
          <th>Direction</th>
          <th>Stop</th>
        </tr>
      </thead>
      <tbody>
        <tr ng-repeat="stop in nextBus.addForm.nearbyStops" ng-click="nextBus.addForm.addStop(stop.tag + '|' + stop.stopTag)">
          <td>{{stop.title}}</td>
          <td>{{stop.directionTitle}}</td>
          <td>{{stop.stopTitle}}</td>
        </tr>
      </tbody>
    </table>

  </div>
</div>
