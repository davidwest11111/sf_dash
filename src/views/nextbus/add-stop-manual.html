<div class="card">
    <div class="card-body">
      <form ng-submit="nextBus.addForm.validate()">
        <div class="row">
          <div class="form-group col-6">
            <label>Route Tag:</label>
            <input type="text" class="form-control" ng-model="nextBus.addForm.routeTag">
          </div>
          <div class="form-group col-6">
            <label>Stop Tag:</label>
            <input type="text" class="form-control" ng-model="nextBus.addForm.stopTag">
          </div>
        </div>
        <div>
          <input class="btn btn-primary" type="submit" value="Validate &amp; Add Stop">
        </div>
      </form>
    </div>
</div>
