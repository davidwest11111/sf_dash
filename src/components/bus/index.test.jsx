import React from 'react';
import { shallow } from 'enzyme';
import shallowToJson from 'enzyme-to-json';
import moment from 'moment';

import { SAMPLE_PREDICTION } from '../../sample_data/nextbus';
import Bus from './index';


describe('outputs the expected tree when', () => {
  let wrapper;
  let defaultProps;

  beforeEach(() => {
    defaultProps = {
      showBusRemoval: false,
      showBusAddStopForm: false,
      lastUpdated: moment(),
      msUntilWarning: 10000,
      removeStopRoute: () => {},
      toggleBusRemove: () => {},
      addForm: {
        toggleBusAddStopForm: () => {},
        loading: false,
        errMsg: '',
        validate: () => {},
        getNearbyStops: () => {},
        addStop: () => {},
        nearbyStops: [
          {
            tag: 'FIXME',
            stopTag: 'FIXME',
            title: 'FIXME',
            directionTitle: 'FIXME',
            stopTitle: 'FIXME',
          },
        ],
      },
    };
  });

  test('there are no predictions', () => {
    wrapper = shallow((
      <Bus {...defaultProps} predictions={[]} />
    ));
  });

  test('there is at least one prediction in the array', () => {
    wrapper = shallow((
      <Bus {...defaultProps} predictions={[{ ...SAMPLE_PREDICTION }]} />
    ));
  });

  afterEach(() => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
