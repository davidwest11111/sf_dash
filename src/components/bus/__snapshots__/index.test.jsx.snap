// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputs the expected tree when there are no predictions 1`] = `
<div
  className="bus"
>
  <Card
    tag="div"
  >
    <CardBody
      tag="div"
    >
      No bus stops added.
    </CardBody>
  </Card>
  <Predictions
    className=""
    predictions={Array []}
    removeStopRoute={[Function]}
  />
  <div
    className="d-flex align-items-start justify-content-between"
  >
    <EditTogglers
      addForm={
        Object {
          "addStop": [Function],
          "errMsg": "",
          "getNearbyStops": [Function],
          "loading": false,
          "nearbyStops": Array [
            Object {
              "directionTitle": "FIXME",
              "stopTag": "FIXME",
              "stopTitle": "FIXME",
              "tag": "FIXME",
              "title": "FIXME",
            },
          ],
          "toggleBusAddStopForm": [Function],
          "validate": [Function],
        }
      }
      className=""
      predictions={Array []}
      showBusRemoval={false}
      toggleBusRemove={[Function]}
    />
    <TimeSinceLastUpdated
      className=""
      lastUpdated={"2019-12-11T00:00:00.000Z"}
      msUntilWarning={10000}
    />
  </div>
</div>
`;

exports[`outputs the expected tree when there is at least one prediction in the array 1`] = `
<div
  className="bus"
>
  <Predictions
    className=""
    predictions={
      Array [
        Object {
          "$$hashKey": "object:17",
          "_agencyTitle": "SF Muni",
          "_routeTag": "14",
          "_routeTitle": "14-Mission",
          "_stopTag": "5565",
          "_stopTitle": "Mission St & 24th St",
          "className": "",
          "direction": Object {
            "_title": "Inbound to Ferry Plaza",
            "prediction": Array [
              Object {
                "$$hashKey": "object:19",
                "_block": "1620",
                "_dirTag": "14___I_F00",
                "_epochTime": "1588198229803",
                "_isDeparture": "false",
                "_minutes": "4",
                "_seconds": "296",
                "_tripTag": "9359239",
                "_vehicle": "6609",
              },
              Object {
                "$$hashKey": "object:20",
                "_block": "1622",
                "_dirTag": "14___I_F00",
                "_epochTime": "1588198681179",
                "_isDeparture": "false",
                "_minutes": "12",
                "_seconds": "747",
                "_tripTag": "9359240",
                "_vehicle": "6599",
              },
              Object {
                "$$hashKey": "object:21",
                "_block": "1633",
                "_dirTag": "14___I_F00",
                "_epochTime": "1588199412858",
                "_isDeparture": "false",
                "_minutes": "24",
                "_seconds": "1479",
                "_tripTag": "9359237",
                "_vehicle": "6547",
              },
              Object {
                "$$hashKey": "object:22",
                "_block": "1628",
                "_dirTag": "14___I_F00",
                "_epochTime": "1588199419316",
                "_isDeparture": "false",
                "_minutes": "24",
                "_seconds": "1486",
                "_tripTag": "9359243",
                "_vehicle": "6533",
              },
            ],
          },
          "message": Array [
            Object {
              "_priority": "Normal",
              "_text": "Service frequency for this stop is 10 minutes or less.",
            },
            Object {
              "_priority": "Low",
              "_text": "Muni for essential trips only. Masks or face coverings are required. Practice physical distancing. Visit sfmta.com/COVID19 for more info.",
            },
            Object {
              "_priority": "Low",
              "_text": "Text COVID19SF to 888-777 for official updates. Visit sfdph.org",
            },
            Object {
              "_priority": "Normal",
              "_text": "Board thru rear doors. Front door for accessibility needs only.",
            },
          ],
        },
      ]
    }
    removeStopRoute={[Function]}
  />
  <div
    className="d-flex align-items-start justify-content-between"
  >
    <EditTogglers
      addForm={
        Object {
          "addStop": [Function],
          "errMsg": "",
          "getNearbyStops": [Function],
          "loading": false,
          "nearbyStops": Array [
            Object {
              "directionTitle": "FIXME",
              "stopTag": "FIXME",
              "stopTitle": "FIXME",
              "tag": "FIXME",
              "title": "FIXME",
            },
          ],
          "toggleBusAddStopForm": [Function],
          "validate": [Function],
        }
      }
      className=""
      predictions={
        Array [
          Object {
            "$$hashKey": "object:17",
            "_agencyTitle": "SF Muni",
            "_routeTag": "14",
            "_routeTitle": "14-Mission",
            "_stopTag": "5565",
            "_stopTitle": "Mission St & 24th St",
            "className": "",
            "direction": Object {
              "_title": "Inbound to Ferry Plaza",
              "prediction": Array [
                Object {
                  "$$hashKey": "object:19",
                  "_block": "1620",
                  "_dirTag": "14___I_F00",
                  "_epochTime": "1588198229803",
                  "_isDeparture": "false",
                  "_minutes": "4",
                  "_seconds": "296",
                  "_tripTag": "9359239",
                  "_vehicle": "6609",
                },
                Object {
                  "$$hashKey": "object:20",
                  "_block": "1622",
                  "_dirTag": "14___I_F00",
                  "_epochTime": "1588198681179",
                  "_isDeparture": "false",
                  "_minutes": "12",
                  "_seconds": "747",
                  "_tripTag": "9359240",
                  "_vehicle": "6599",
                },
                Object {
                  "$$hashKey": "object:21",
                  "_block": "1633",
                  "_dirTag": "14___I_F00",
                  "_epochTime": "1588199412858",
                  "_isDeparture": "false",
                  "_minutes": "24",
                  "_seconds": "1479",
                  "_tripTag": "9359237",
                  "_vehicle": "6547",
                },
                Object {
                  "$$hashKey": "object:22",
                  "_block": "1628",
                  "_dirTag": "14___I_F00",
                  "_epochTime": "1588199419316",
                  "_isDeparture": "false",
                  "_minutes": "24",
                  "_seconds": "1486",
                  "_tripTag": "9359243",
                  "_vehicle": "6533",
                },
              ],
            },
            "message": Array [
              Object {
                "_priority": "Normal",
                "_text": "Service frequency for this stop is 10 minutes or less.",
              },
              Object {
                "_priority": "Low",
                "_text": "Muni for essential trips only. Masks or face coverings are required. Practice physical distancing. Visit sfmta.com/COVID19 for more info.",
              },
              Object {
                "_priority": "Low",
                "_text": "Text COVID19SF to 888-777 for official updates. Visit sfdph.org",
              },
              Object {
                "_priority": "Normal",
                "_text": "Board thru rear doors. Front door for accessibility needs only.",
              },
            ],
          },
        ]
      }
      showBusRemoval={false}
      toggleBusRemove={[Function]}
    />
    <TimeSinceLastUpdated
      className=""
      lastUpdated={"2019-12-11T00:00:00.000Z"}
      msUntilWarning={10000}
    />
  </div>
</div>
`;
